File: js/src/stream/cursor.ts
/**
 * StreamCursor - Per-SID state tracking for GS1 streams.
 */

import { Frame, FrameKind, BaseMismatchError } from './types';
import { verifyBase, stateHashLooseSync } from './hash';
import { GValue } from '../types';

/**
 * State for a single stream ID.
 */
export interface SIDState {
  sid: bigint;
  lastSeq: bigint;
  lastAcked: bigint;
  stateHash: Uint8Array | null;
  state: GValue | null;
  final: boolean;
}

/**
 * StreamCursor tracks per-SID state for stream processing.
 */
export class StreamCursor {
  private cursors = new Map<bigint, SIDState>();

  /**
   * Get state for a SID, creating it if needed.
   */
  get(sid: bigint): SIDState {
    let state = this.cursors.get(sid);
    if (!state) {
      state = {
        sid,
        lastSeq: 0n,
        lastAcked: 0n,
        stateHash: null,
        state: null,
        final: false,
      };
      this.cursors.set(sid, state);
    }
    return state;
  }

  /**
   * Get state for a SID without creating it.
   */
  getReadOnly(sid: bigint): SIDState | undefined {
    return this.cursors.get(sid);
  }

  /**
   * Delete state for a SID.
   */
  delete(sid: bigint): void {
    this.cursors.delete(sid);
  }

  /**
   * Get all tracked SIDs.
   */
  allSIDs(): bigint[] {
    return Array.from(this.cursors.keys());
  }

  /**
   * Process a frame and update cursor state.
   * Throws on sequence gaps, duplicates, or base mismatches.
   */
  processFrame(frame: Frame): void {
    const state = this.get(frame.sid);

    // Check sequence monotonicity
    if (frame.seq !== 0n && frame.seq <= state.lastSeq) {
      throw new Error(`sequence not monotonic: got ${frame.seq}, last was ${state.lastSeq}`);
    }

    // Check for gaps
    if (state.lastSeq > 0n && frame.seq !== state.lastSeq + 1n) {
      throw new Error(`sequence gap: expected ${state.lastSeq + 1n}, got ${frame.seq}`);
    }

    // For patches with base, verify state hash
    if (frame.kind === 'patch' && frame.base && state.stateHash) {
      if (!verifyBase(state.stateHash, frame.base)) {
        throw new BaseMismatchError();
      }
    }

    // Update sequence
    state.lastSeq = frame.seq;

    // Update final flag
    if (frame.final) {
      state.final = true;
    }
  }

  /**
   * Set the current state and compute its hash.
   */
  setState(sid: bigint, value: GValue): void {
    const state = this.get(sid);
    state.state = value;
    state.stateHash = stateHashLooseSync(value);
  }

  /**
   * Set the state hash directly.
   */
  setStateHash(sid: bigint, hash: Uint8Array): void {
    const state = this.get(sid);
    state.stateHash = hash;
  }

  /**
   * Mark a sequence as acknowledged.
   */
  ack(sid: bigint, seq: bigint): void {
    const state = this.get(sid);
    if (seq > state.lastAcked) {
      state.lastAcked = seq;
    }
  }

  /**
   * Get sequences that have been seen but not acked.
   */
  pendingAcks(sid: bigint): bigint[] {
    const state = this.getReadOnly(sid);
    if (!state || state.lastSeq <= state.lastAcked) {
      return [];
    }

    const pending: bigint[] = [];
    for (let seq = state.lastAcked + 1n; seq <= state.lastSeq; seq++) {
      pending.push(seq);
    }
    return pending;
  }

  /**
   * Check if resync is needed (no state hash).
   */
  needsResync(sid: bigint): boolean {
    const state = this.getReadOnly(sid);
    return !state || !state.stateHash;
  }
}

/**
 * Frame handler callbacks.
 */
export interface FrameHandlerCallbacks {
  onDoc?: (sid: bigint, seq: bigint, payload: Uint8Array, state: SIDState) => void;
  onPatch?: (sid: bigint, seq: bigint, payload: Uint8Array, state: SIDState) => void;
  onRow?: (sid: bigint, seq: bigint, payload: Uint8Array, state: SIDState) => void;
  onUI?: (sid: bigint, seq: bigint, payload: Uint8Array, state: SIDState) => void;
  onAck?: (sid: bigint, seq: bigint, state: SIDState) => void;
  onErr?: (sid: bigint, seq: bigint, payload: Uint8Array, state: SIDState) => void;
  onFinal?: (sid: bigint, state: SIDState) => void;
  onSeqGap?: (sid: bigint, expected: bigint, got: bigint) => boolean; // Return true to allow
  onBaseMismatch?: (sid: bigint, frame: Frame) => boolean; // Return true to allow
}

/**
 * FrameHandler processes frames with state tracking and callbacks.
 */
export class FrameHandler {
  readonly cursor: StreamCursor;
  private callbacks: FrameHandlerCallbacks;

  constructor(callbacks: FrameHandlerCallbacks = {}) {
    this.cursor = new StreamCursor();
    this.callbacks = callbacks;
  }

  /**
   * Handle a frame and call the appropriate callback.
   */
  handle(frame: Frame): void {
    const state = this.cursor.get(frame.sid);

    // Check sequence
    if (frame.seq !== 0n && state.lastSeq > 0n) {
      if (frame.seq <= state.lastSeq) {
        // Duplicate - skip
        return;
      }
      if (frame.seq !== state.lastSeq + 1n) {
        // Gap detected
        if (this.callbacks.onSeqGap) {
          const allow = this.callbacks.onSeqGap(frame.sid, state.lastSeq + 1n, frame.seq);
          if (!allow) return;
        }
      }
    }

    // Check base for patches
    if (frame.kind === 'patch' && frame.base && state.stateHash) {
      if (!verifyBase(state.stateHash, frame.base)) {
        if (this.callbacks.onBaseMismatch) {
          const allow = this.callbacks.onBaseMismatch(frame.sid, frame);
          if (!allow) return;
        } else {
          throw new BaseMismatchError();
        }
      }
    }

    // Update sequence
    state.lastSeq = frame.seq;

    // Dispatch to callback
    switch (frame.kind) {
      case 'doc':
        this.callbacks.onDoc?.(frame.sid, frame.seq, frame.payload, state);
        break;
      case 'patch':
        this.callbacks.onPatch?.(frame.sid, frame.seq, frame.payload, state);
        break;
      case 'row':
        this.callbacks.onRow?.(frame.sid, frame.seq, frame.payload, state);
        break;
      case 'ui':
        this.callbacks.onUI?.(frame.sid, frame.seq, frame.payload, state);
        break;
      case 'ack':
        this.callbacks.onAck?.(frame.sid, frame.seq, state);
        break;
      case 'err':
        this.callbacks.onErr?.(frame.sid, frame.seq, frame.payload, state);
        break;
    }

    // Handle final
    if (frame.final) {
      state.final = true;
      this.callbacks.onFinal?.(frame.sid, state);
    }
  }
}

File: go/stream/cursor.go
package stream

import (
	"fmt"
	"sync"

	"github.com/Neumenon/glyph/glyph"
)

// StreamCursor tracks per-SID state for stream processing.
// It maintains sequence numbers, state hashes, and provides
// helpers for patch verification and acknowledgement.
type StreamCursor struct {
	mu sync.RWMutex

	// Per-SID state
	cursors map[uint64]*SIDState
}

// SIDState holds state for a single stream ID.
type SIDState struct {
	SID       uint64
	LastSeq   uint64        // Last sequence number seen
	LastAcked uint64        // Last sequence number acknowledged
	StateHash [32]byte      // Current state hash (for patch verification)
	HasState  bool          // Whether StateHash is valid
	State     *glyph.GValue // Current state document (optional)
	Final     bool          // Whether stream has ended
}

// NewStreamCursor creates a new stream cursor.
func NewStreamCursor() *StreamCursor {
	return &StreamCursor{
		cursors: make(map[uint64]*SIDState),
	}
}

// Get returns the state for a SID, creating it if needed.
func (sc *StreamCursor) Get(sid uint64) *SIDState {
	sc.mu.Lock()
	defer sc.mu.Unlock()

	state, ok := sc.cursors[sid]
	if !ok {
		state = &SIDState{SID: sid}
		sc.cursors[sid] = state
	}
	return state
}

// GetReadOnly returns the state for a SID without creating it.
func (sc *StreamCursor) GetReadOnly(sid uint64) *SIDState {
	sc.mu.RLock()
	defer sc.mu.RUnlock()
	return sc.cursors[sid]
}

// Delete removes state for a SID.
func (sc *StreamCursor) Delete(sid uint64) {
	sc.mu.Lock()
	defer sc.mu.Unlock()
	delete(sc.cursors, sid)
}

// AllSIDs returns all tracked SIDs.
func (sc *StreamCursor) AllSIDs() []uint64 {
	sc.mu.RLock()
	defer sc.mu.RUnlock()

	sids := make([]uint64, 0, len(sc.cursors))
	for sid := range sc.cursors {
		sids = append(sids, sid)
	}
	return sids
}

// ProcessFrame processes a frame and updates cursor state.
// Returns an error if:
//   - Sequence number is not monotonic (gap or duplicate)
//   - Base hash mismatch for patch frames
//
// On success, updates LastSeq and returns nil.
func (sc *StreamCursor) ProcessFrame(frame *Frame) error {
	state := sc.Get(frame.SID)

	// Check sequence monotonicity
	if frame.Seq != 0 && frame.Seq <= state.LastSeq {
		return fmt.Errorf("sequence not monotonic: got %d, last was %d", frame.Seq, state.LastSeq)
	}

	// Check for gaps
	if state.LastSeq > 0 && frame.Seq != state.LastSeq+1 {
		return fmt.Errorf("sequence gap: expected %d, got %d", state.LastSeq+1, frame.Seq)
	}

	// For patches with base, verify state hash
	if frame.Kind == KindPatch && frame.Base != nil {
		if !state.HasState {
			return fmt.Errorf("cannot verify base: no state hash for SID %d", frame.SID)
		}
		if !VerifyBase(state.StateHash, *frame.Base) {
			return &BaseMismatchError{Expected: *frame.Base, Got: state.StateHash}
		}
	}

	// Update sequence
	state.LastSeq = frame.Seq

	// Update final flag
	if frame.IsFinal() {
		state.Final = true
	}

	return nil
}

// SetState sets the current state and computes its hash.
// Use this after applying a doc snapshot or patch.
func (sc *StreamCursor) SetState(sid uint64, value *glyph.GValue) {
	state := sc.Get(sid)
	state.State = value
	state.StateHash = StateHashLoose(value)
	state.HasState = true
}

// SetStateHash sets the state hash directly.
// Use this when you have pre-computed the hash.
func (sc *StreamCursor) SetStateHash(sid uint64, hash [32]byte) {
	state := sc.Get(sid)
	state.StateHash = hash
	state.HasState = true
}

// Ack marks a sequence as acknowledged.
func (sc *StreamCursor) Ack(sid, seq uint64) {
	state := sc.Get(sid)
	if seq > state.LastAcked {
		state.LastAcked = seq
	}
}

// PendingAcks returns sequences that have been seen but not acked.
func (sc *StreamCursor) PendingAcks(sid uint64) []uint64 {
	state := sc.GetReadOnly(sid)
	if state == nil {
		return nil
	}

	if state.LastSeq <= state.LastAcked {
		return nil
	}

	pending := make([]uint64, 0, state.LastSeq-state.LastAcked)
	for seq := state.LastAcked + 1; seq <= state.LastSeq; seq++ {
		pending = append(pending, seq)
	}
	return pending
}

// NeedsResync returns true if there's a gap or state mismatch.
func (sc *StreamCursor) NeedsResync(sid uint64) bool {
	state := sc.GetReadOnly(sid)
	if state == nil {
		return true
	}
	return !state.HasState
}

// ============================================================
// Frame Handler - functional processing helper
// ============================================================

// FrameHandler processes frames with state tracking.
type FrameHandler struct {
	Cursor *StreamCursor

	// Callbacks (optional)
	OnDoc   func(sid uint64, seq uint64, payload []byte, state *SIDState) error
	OnPatch func(sid uint64, seq uint64, payload []byte, state *SIDState) error
	OnRow   func(sid uint64, seq uint64, payload []byte, state *SIDState) error
	OnUI    func(sid uint64, seq uint64, payload []byte, state *SIDState) error
	OnAck   func(sid uint64, seq uint64, state *SIDState) error
	OnErr   func(sid uint64, seq uint64, payload []byte, state *SIDState) error
	OnFinal func(sid uint64, state *SIDState) error

	// Error handling
	OnSeqGap       func(sid uint64, expected, got uint64) error // Called on sequence gap
	OnBaseMismatch func(sid uint64, frame *Frame) error         // Called on base hash mismatch
}

// NewFrameHandler creates a handler with default cursor.
func NewFrameHandler() *FrameHandler {
	return &FrameHandler{
		Cursor: NewStreamCursor(),
	}
}

// Handle processes a frame and calls the appropriate callback.
func (h *FrameHandler) Handle(frame *Frame) error {
	state := h.Cursor.Get(frame.SID)

	// Check sequence
	if frame.Seq != 0 && state.LastSeq > 0 {
		if frame.Seq <= state.LastSeq {
			// Duplicate or out of order - skip
			return nil
		}
		if frame.Seq != state.LastSeq+1 {
			// Gap detected
			if h.OnSeqGap != nil {
				if err := h.OnSeqGap(frame.SID, state.LastSeq+1, frame.Seq); err != nil {
					return err
				}
			}
		}
	}

	// Check base for patches
	if frame.Kind == KindPatch && frame.Base != nil && state.HasState {
		if !VerifyBase(state.StateHash, *frame.Base) {
			if h.OnBaseMismatch != nil {
				return h.OnBaseMismatch(frame.SID, frame)
			}
			return &BaseMismatchError{Expected: *frame.Base, Got: state.StateHash}
		}
	}

	// Update sequence
	state.LastSeq = frame.Seq

	// Dispatch to callback
	var err error
	switch frame.Kind {
	case KindDoc:
		if h.OnDoc != nil {
			err = h.OnDoc(frame.SID, frame.Seq, frame.Payload, state)
		}
	case KindPatch:
		if h.OnPatch != nil {
			err = h.OnPatch(frame.SID, frame.Seq, frame.Payload, state)
		}
	case KindRow:
		if h.OnRow != nil {
			err = h.OnRow(frame.SID, frame.Seq, frame.Payload, state)
		}
	case KindUI:
		if h.OnUI != nil {
			err = h.OnUI(frame.SID, frame.Seq, frame.Payload, state)
		}
	case KindAck:
		if h.OnAck != nil {
			err = h.OnAck(frame.SID, frame.Seq, state)
		}
	case KindErr:
		if h.OnErr != nil {
			err = h.OnErr(frame.SID, frame.Seq, frame.Payload, state)
		}
	}

	if err != nil {
		return err
	}

	// Handle final
	if frame.IsFinal() {
		state.Final = true
		if h.OnFinal != nil {
			return h.OnFinal(frame.SID, state)
		}
	}

	return nil
}
